## Task 1
Define a variable my_sent to be a list of words <br>

Use syntax <br>
`my_sent = ["My", "sent"]` (but with your own words, or a favorite saying). <br>

Use `' '.join(my_sent)` to convert this into a string <br>
Use `split()` to split the string back into the list form you had to start with.


In [3]:
my_sent = ["There", "is", "a", "simple", "way", "to", "place", "text", "which", "will", "not", "be", "modified", "in", "Jupyter", "Notebook "]
my_sent = " ".join(my_sent)
my_sent

'There is a simple way to place text which will not be modified in Jupyter Notebook '

In [4]:
my_sent = my_sent.split()
print(my_sent)

['There', 'is', 'a', 'simple', 'way', 'to', 'place', 'text', 'which', 'will', 'not', 'be', 'modified', 'in', 'Jupyter', 'Notebook']


## Task 2

Define several variables containing lists of words, e.g., **phrase1, phrase2** and so on. Join them together in various combinations **(using the plus operator)** to form whole sentences. <br> <br> What is the relationship between len(phrase1 + phrase2) and len(phrase1) + len(phrase2)?



In [5]:
phrase1 = ["Rainy","days","are","about","to","come"]
phrase2 = ["Nowadays","youths","are","active","in","politics"]
phrase3 = ["Programming","is","all","about","practise"]
phrase4 = ["I","like","travelling"]

In [6]:
combination1 = ' '.join(phrase1+phrase2)
combination2 = ' '.join(phrase2+phrase1+phrase3)
combination3 = ' '.join(phrase2+phrase1+phrase4+phrase3)

In [7]:
print(combination1)
print(combination2)
print(combination3)

Rainy days are about to come Nowadays youths are active in politics
Nowadays youths are active in politics Rainy days are about to come Programming is all about practise
Nowadays youths are active in politics Rainy days are about to come I like travelling Programming is all about practise


In [8]:
len(phrase1 + phrase2)  == len(phrase1) + len(phrase2)

True

## Task 3
What is the difference between the following two lines? Which one will give a larger value? Will this be the case for other texts?<br>
* `sorted(set(w.lower() for w in text1))`
* `sorted(w.lower() for w in set(text1))`


In [9]:

sorted(set(w.lower() for w in phrase1))

['about', 'are', 'come', 'days', 'rainy', 'to']

In [10]:
sorted(w.lower() for w in set(phrase1))

['about', 'are', 'come', 'days', 'rainy', 'to']

## Task 4
What is the difference between the following two tests? <br>
* `w.isupper()`
* `not w.islower()`


 **Answer**
 * `w.isupper()` checks whether all characters are in uppercase in a word
 * `not w.islower()` checks whether there is at least one uppercase character  in a word.



In [11]:
word = 'Water'
word.isupper()

False

In [12]:
not word.islower()

True

Thus, we can see that `word.isupper()` returns `False` due to presence of one uppercase letter 
<br>
`not w.islower()` returns `True` as one uppercase letter is present

# Task 5

What does the following Python code do?
<br>
`sum(len(w) for w in text1)`
<br>
Can you use it to work out the average word length of a text?


## Answer

`sum(len(w) for w in text1)` gives the  total number of characters present in the  `text1`.


In [13]:
#when entire text is string space count is also included

text1= 'Define several variables containing lists of words and join them together in various combinations'
sum(len(w) for w in text1)

97

In [14]:
#when text is splitted into a list, sum of character of strings are counted excluding space

text = text1.split()
sum(len(w) for w in text)

84

## Calculation of average word length of a text

In [15]:
text = text1.split()
avg_word_len= sum(len(w) for w in text)/len(text)
print("Average word length is ", avg_word_len)

Average word length is  6.0


# Task 6


Define a function `percent(word, text)` that calculates how often a given word occurs in a text, and expresses the result as a percentage.

In [16]:
def percent(word,text):
    text= text.split()
    frequency= text.count(word)
    total_words= len(text)
    percentage= (frequency/total_words)*100
    return f' {word} covers {percentage} % of the total text'

In [17]:
percent("home","I go home twice a year but my brother goes home quite frequently")

' home covers 15.384615384615385 % of the total text'

# Task 7

We have been using sets to store vocabularies. Try the following Python expression: set(sent3) < set(text1). Experiment with this using different arguments to set(). What does it do? Can you think of a practical application for this?

In [18]:
sent3= "I go home twice a year but my brother goes home quite frequently"
text1= 'Define several variables containing lists of words and join them together in various combinations'


In [19]:
def compare_unique_word(sent,text1):
    return set(sent) < set(text1)

In [20]:
compare_unique_word(sent3,text1)

False

Set operation returns the unique value out of given iterable . So the given `set(sent) < set(text1)` expression can be used to check if `sent` is subset of `text1` or not.If its true it returns boolean value `True` else `False`

## Task 8

Create a object called translate which you could look up using words in both German and Spanish in order to get corresponding words in English. What problem might arise with this approach? Can you suggest a way to avoid this problem?

### Swadesh 

Swadesh wordlists, in NLTK consists  of about 200 common words in several languages. The languages are identified using an ISO 639 two-letter code.

`['be', 'bg', 'bs', 'ca', 'cs', 'cu', 'de', 'en', 'es', 'fr', 'hr','it',
  'la', 'mk', 'nl', 'pl', 'pt', 'ro', 'ru', 'sk', 'sl', 'sr', 'sw', 'uk']`

In [21]:
from nltk.corpus import swadesh


In [22]:
translate = dict()
german_to_eng = swadesh.entries(['de','en'])
spanish_to_eng = swadesh.entries(['es','en'])
translate.update(dict(german_to_eng))
translate.update(dict(spanish_to_eng))
print(len(translate))

412


Here, german and spanish are both translated to english lanuage so keys will be unique but values for different keys can be same.412 entries are created in translate object.

In [23]:
translate = dict()
eng_to_german = swadesh.entries(['en','de'])
eng_to_spanish = swadesh.entries(['en','es'])
translate.update(dict(eng_to_german))
translate.update(dict(eng_to_spanish))
print(len(translate))


207


Here, english is translated to german in first step and dictionary is updated and in next step we translate english to german, so keys can be same for second step  and  the result of first step of translation. As a result only 207 entries are created in translate object. So different objects are preferable.

## Task 9

Write a program to find all words that occur at least three times in the **Brown Corpus.**


In [38]:

import nltk
nltk.download('brown')

In [35]:
from nltk.corpus import brown

In [36]:
def filter_word_by_freq(frequency):
    unique_words= set(brown.words())
    brown_freq_dist= nltk.FreqDist(brown.words())
    words_of_given_freq= [word for word in unique_words if brown_freq_dist[word]>frequency]
    return words_of_given_freq

In [40]:
words_of_given_freq= filter_word_by_freq(3)


## Task 10

Write a function that finds the 50 most frequently occurring words of a text that are not stopwords.

In [41]:
from nltk.corpus import stopwords

def most_frequent_words(number):
    freq_distribution=nltk.FreqDist(brown.words(categories='news'))
    words= [word for word in freq_distribution]
    words= [word.lower() for word in freq_distribution if word.lower()  not in stopwords.words('english') and  word.isalpha()]
    return words[:number]


In [42]:
frequent_words = most_frequent_words(50)

## Task 11


Write a program to print the 50 most frequent **bigrams** (pairs of adjacent words) of a text, omitting bigrams that contain stopwords

In [43]:
def n_freq_bigrams(n):
    brown_words=brown.words(categories='news')
    words= [word for word in brown_words]
    words_without_stopwords= [word.lower() for word in words if word.lower()  not in stopwords.words('english') and  word.isalpha()]
    bigrams = [(a,b) for a,b in nltk.bigrams(words_without_stopwords) ]
    bigrams_fd = nltk.FreqDist(bigrams)
    return bigrams_fd.most_common(n)

In [44]:
print(n_freq_bigrams(50))

[(('new', 'york'), 52), (('per', 'cent'), 50), (('united', 'states'), 37), (('last', 'week'), 35), (('last', 'year'), 34), (('white', 'house'), 29), (('high', 'school'), 23), (('home', 'runs'), 23), (('president', 'kennedy'), 20), (('last', 'night'), 18), (('said', 'would'), 15), (('years', 'ago'), 15), (('san', 'francisco'), 15), (('premier', 'khrushchev'), 13), (('kansas', 'city'), 13), (('vice', 'president'), 12), (('los', 'angeles'), 12), (('united', 'nations'), 11), (('new', 'orleans'), 11), (('police', 'said'), 11), (('sales', 'tax'), 10), (('two', 'years'), 10), (('american', 'catholic'), 10), (('mantle', 'maris'), 10), (('jury', 'said'), 9), (('would', 'like'), 9), (('el', 'paso'), 9), (('social', 'security'), 9), (('kennedy', 'administration'), 9), (('first', 'time'), 9), (('country', 'club'), 9), (('rules', 'committee'), 9), (('air', 'force'), 9), (('tax', 'bill'), 9), (('grand', 'jury'), 8), (('city', 'council'), 8), (('rhode', 'island'), 8), (('president', 'said'), 8), (('w

## Task 12

Write a program to create a table of word frequencies by genre, like the one given in 1 for modals. Choose your own words and try to find words whose presence (or absence) is typical of a genre. Discuss your findings

In [55]:
import nltk
cfd = nltk.ConditionalFreqDist((genre,word)for genre in brown.categories() for word in brown.words(categories=genre))
# general_words = ["love", "hate", "death", "life", "marriage", "work", "children","magic"]
general_words = ["comedy", "ego", "death","money", "dance", "night", "marriage", "calm", "parents","politics"]

cfd.tabulate(conditions=brown.categories(), samples=general_words)


                  comedy      ego    death    money    dance    night marriage     calm  parents politics 
      adventure        0        0       20       16        7       30        4        6        1        1 
 belles_lettres        3        4       57       39       16       44       12        6        8       16 
      editorial        1        1       21       15        4       16        4        3        6        7 
        fiction        0        1       17       19        2       53        1        6        7        4 
     government        0        0        2       13        3        8        4        0        0        4 
        hobbies        0        1        3       16       10        2        0        0        7        0 
          humor       14        2        1        3        1       11        2        0        1        1 
        learned        1        3       21       10        0        7       14        3       14       12 
           lore        3        1    

Comedy has no reference to adventure, mystery, science_fiction and romance. While belles_letter has more concern with words like death, money, night.